### PR TITLE
feat: Show diff UI when external server data changes

### PR DESCRIPTION
**Pull request by ghoshkaj**
**Note: This PR is targeting a shared dev branch, not main. These changes are strictly a hack for prototyping, and will not be merged into main 🙂**

This PR does a number of things:
1. It creates a different form to simulate the external data server
2. It triggers and completes the full loop of using a signal to deliver the information that the external data has changed
3. It introduces elementary conflict resolution UI to demonstrate the behaviour on the client app when the external server changes it's data.

Here is what that full loop looks like currently:

full_flow_with_ignore_changes

### CODE REVIEW

File: examples/hosts/app-integration/external-data/src/model-interface/index.ts [hunk 1]
```diff
@@ -46,7 +39,12 @@ export interface ITaskEvents extends IEvent {
     /**
      * Emitted when the name or priority have changed respectively.
      */
-    (event: "nameChanged" | "priorityChanged", listener: () => void);
+    (event:
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/src/model-interface/index.ts:42]: I know you said this is a hack, but nit here, probably preferrable to have each listener be declared separately even if the signature for all of them is the same. I personally find it cleaner, and I think there was some advantage for automated documentation... @Josmithr I think it was you pointing that out in other PRs I've seen in the past?

File: examples/hosts/app-integration/external-data/src/model/containerCode.ts [hunk 2]
```diff
@@ -62,11 +62,12 @@ export class TaskListContainerRuntimeFactory extends ModelContainerRuntimeFactor
         );
         // Register listener only once the model is fully loaded and ready
         runtime.on("signal", (message) => {
-            if (message.type === SignalType.ExternalDataChanged) {
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+        if ('content' in message && 'type' in message.content && message.content.type === SignalType.ExternalDataChanged) {
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/src/model/containerCode.ts:66]:
```suggestion
        if (message?.content?.type === SignalType.ExternalDataChanged) {
```
Can it be done this way?

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx [hunk 3]
```diff
@@ -5,17 +5,14 @@
 
 import React, { useEffect, useState } from "react";
 import isEqual from 'lodash.isequal'
-import type { IAppModel, TaskData } from "../model-interface";
-import { customerServicePort } from "../mock-service-interface";
+import type { TaskData } from "../model-interface";
+import { customerServicePort  } from "../mock-service-interface";
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/src/view/debugView.tsx:9]:
```suggestion
import { customerServicePort } from "../mock-service-interface";
```

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx [hunk 4]
```diff
@@ -158,14 +156,151 @@ function debugResetExternalData(): void {
...
+                    setExternalData(newData);
+                }
+            } catch(error) {
+                console.error(`APP: An error was encountered while polling external data:\n${error}`);
+            }
+        }
+
+        // Run once immediately to run without waiting.
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        pollForServiceUpdates();
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/src/view/debugView.tsx:244]:
```suggestion
        pollForServiceUpdates().catch(console.error);
```
Nit. So failures during the request at least get printed somewhere, and I think this should make the eslint-disable unnecessary.

**Review by alexvy86 — COMMENTED** [review]
Getting familiar with React little by little but I make no guarantees that my review of that part was exhaustive 😅

File: experimental/framework/react-inputs/src/CollaborativeInput.tsx [hunk 5]
```diff
@@ -57,9 +57,9 @@ export class CollaborativeInput
     public componentDidMount() {
         // Sets an event listener so we can update our state as the value changes
         this.props.sharedString.on("sequenceDelta", (ev: SequenceDeltaEvent) => {
-            if (!ev.isLocal) {
+            // if (!ev.isLocal) {
```
Comment (Josmithr) [inline experimental/framework/react-inputs/src/CollaborativeInput.tsx:60]: Why was this change necessary? If the event is local, then I believe we should already be updating our local state value.

File: examples/hosts/app-integration/external-data/src/view/taskListView.tsx [hunk 6]
```diff
@@ -20,16 +20,33 @@ interface ITaskRowProps {
...
+            console.log(savedName);
+            console.log(savedDiffType);
+        }
         task.on("priorityChanged", updateFromRemotePriority);
+        task.on("externalPriorityChanged", showSavedPriority);
+        task.on("externalNameChanged", showSavedName);
         updateFromRemotePriority();
         return (): void => {
             task.off("priorityChanged", updateFromRemotePriority);
+            task.on("externalPriorityChanged", showSavedPriority);
```
Comment (Josmithr) [inline examples/hosts/app-integration/external-data/src/view/taskListView.tsx:48]: These should be "off" calls

File: experimental/framework/react-inputs/src/CollaborativeInput.tsx (hunk 5 above)
Comment (ghoshkaj) [reply experimental/framework/react-inputs/src/CollaborativeInput.tsx:60]: We don't render the change to the screen without removing this `isLocal`. I forget what the orignal logic was for putting this in -- @ChumpChief I think you mentioned it in our conversation last week? Do you remember?

File: examples/hosts/app-integration/external-data/src/view/taskListView.tsx [hunk 7]
```diff
@@ -20,16 +20,33 @@ interface ITaskRowProps {
...
+            task.on("externalNameChanged", showSavedName);
         };
     }, [task]);
```
Comment (Josmithr) [inline examples/hosts/app-integration/external-data/src/view/taskListView.tsx:51]: I believe we need to update the effect deps here to reflect the new data you're using.

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx [hunk 8]
```diff
@@ -158,14 +156,151 @@ function debugResetExternalData(): void {
...
                 <button onClick={ debugResetExternalData }>Reset external data</button><br />
-                <button onClick={ props.model.debugSendCustomSignal }>Trigger external data change signal</button><br />
             </div>
         </div>
     );
 };
+
+
+interface ITaskRowProps {
+    task: ExternalDataTask
```
Comment (Josmithr) [inline examples/hosts/app-integration/external-data/src/view/debugView.tsx:171]: Nit
```suggestion
    task: ExternalDataTask;
```

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx [hunk 9]
```diff
@@ -158,14 +156,151 @@ function debugResetExternalData(): void {
...
+}
+
+/**
+ * The view for a single task in the TaskListView, as a table row.
+ */
+const TaskRow: React.FC<ITaskRowProps> = (props: ITaskRowProps) => {
+    const { task } = props;
+
+    const IdChangeHandler = (e: React.SyntheticEvent<HTMLInputElement>): void => {
```
Comment (Josmithr) [inline examples/hosts/app-integration/external-data/src/view/debugView.tsx:180]: Nit: I think the handlers want to be named in `camelCase`, since they're non-React-component functions.

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx [hunk 10]
```diff
@@ -28,13 +25,14 @@ export interface IDebugViewProps {
  *
  * For the purposes of this test app, it is useful to be able to see both data sources side-by-side.
  */
-export const DebugView: React.FC<IDebugViewProps> = (props: IDebugViewProps) => {
+export const DebugView: React.FC<IDebugViewProps> = () => {
```
Comment (Josmithr) [inline examples/hosts/app-integration/external-data/src/view/debugView.tsx:28]: (Not for this PR) I wonder if we should rename this to "ExternalDataServiceView" or something. "DebugView" isn't super descriptive.

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx [hunk 11]
```diff
@@ -158,14 +156,151 @@ function debugResetExternalData(): void {
...
+    return (
+        <tr>
+            <td><input defaultValue={ task.id }  style={{ width: "30px" }} onChange={ IdChangeHandler }></input></td>
+            <td><input defaultValue={ task.name } style={{ width: "200px" }} onChange={ NameChangeHandler }></input></td>
+            <td><input defaultValue={ task.priority } type="number" style={{ width: "50px" }} onChange={ PriorityChangeHandler }></input></td>
+        </tr>
+    );
+};
+
+class ExternalDataTask {
```
Comment (Josmithr) [inline examples/hosts/app-integration/external-data/src/view/debugView.tsx:199]: Nit: Does this need to be a class? Or can it just be an interface?

Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/view/debugView.tsx:199]: I turned it into a class so that I could pass it into the TaskRow component and manipulate the values from the TaskRow component. Can I do that if it were only an interface?

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx (hunk 10 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/view/debugView.tsx:28]: I agree on both -- we need to rename this as well as maybe do it in a different PR :)

File: examples/hosts/app-integration/external-data/src/model-interface/index.ts (hunk 1 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/model-interface/index.ts:42]: How do I declare each listener separately? As far as I can tell, this is the pattern we've been using in this repo, so I've just stuck to is. Is there an example you can link me to that declares each listener separately? :)

Comment (alexvy86) [reply examples/hosts/app-integration/external-data/src/model-interface/index.ts:42]: Just split each event into its own full line instead of using the union operator for the values of `event`, you can see an example [here](https://github.com/microsoft/FluidFramework/blob/main/packages/common/container-definitions/src/loader.ts#L151-L174).

Comment (alexvy86) [reply examples/hosts/app-integration/external-data/src/model-interface/index.ts:42]: Oh, and the example uses `/* eslint-disable @typescript-eslint/unified-signatures */` and a corresponding `eslint-enable` after that interface declaration because eslint prefers them unified. And I found where Josh [suggested](https://teams.microsoft.com/l/message/19:50292e8934024fc19d6ca2080dd7681e@thread.skype/1666116316603?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=9ce27575-2f82-4689-abdb-bcff07e8063b&parentMessageId=1666116316603&teamName=Fluid%20Framework&channelName=Dev&createdTime=1666116316603&allowXTenantAccess=false) we make each listener its own thing, mostly for documentation purposes, which I agree with.

File: examples/hosts/app-integration/external-data/src/view/taskListView.tsx (hunk 7 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/view/taskListView.tsx]: Is this what you mean here? https://github.com/microsoft/FluidFramework/pull/13732/commits/0467fc9176871cb97399083c19a6a24f75c1a4b2

File: examples/hosts/app-integration/external-data/src/model-interface/index.ts [hunk 12]
```diff
@@ -66,6 +64,38 @@ export interface ITask extends IEventProvider<ITaskEvents> {
...
+     */
+    readonly acceptChange:() => Promise<void>;
+    /**
+     * Ignore the proposed changes to SavedData.
+     */
+    readonly ignoreChange:() => Promise<void>;
+
+    readonly DEFAULT_PRIORITY: number;
+    readonly DEFAULT_NAME: string;
+    readonly DEFAULT_DIFF_TYPE: string;
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/src/model-interface/index.ts:106]: Any reason these are all in caps? I'm used to seeing constants written like that so it confused me for a sec.

File: examples/hosts/app-integration/external-data/src/model/taskList.ts [hunk 13]
```diff
@@ -31,10 +31,19 @@ class Task extends TypedEventEmitter<ITaskEvents> implements ITask {
...
+    public DEFAULT_NAME: string = "";
+    public DEFAULT_DIFF_TYPE: string = "none";
     public constructor(
         private readonly _id: string,
         private readonly _name: SharedString,
-        private readonly _priority: ISharedCell<number>
+        private readonly _priority: ISharedCell<number>,
+        diffName: string = "",
+        diffPriority: number = Number.POSITIVE_INFINITY,
+        diffType: string = "none",
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/src/model/taskList.ts:46]: Can we declare these as `public diffaname: string = ""`, etc, to avoid the declaration in lines 34-36 and the initialization in 55-57?

File: examples/hosts/app-integration/external-data/src/model/taskList.ts [hunk 14]
```diff
@@ -43,6 +52,44 @@ class Task extends TypedEventEmitter<ITaskEvents> implements ITask {
...
+    public externalNameChanged = async (savedName: string): Promise<void> => {
+        this.diffName = savedName;
+        this.diffType = "change";
+        this.emit('externalNameChanged');
+    }
+    public externalPriorityChanged = async (savedPriority: number): Promise<void> => {
+        this.diffPriority = savedPriority;
+        this.diffType = "change";
+        this.emit('externalPriorityChanged');
+    }
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/src/model/taskList.ts:68]: Do these need to be Promises?

File: examples/hosts/app-integration/external-data/src/model/taskList.ts [hunk 15]
```diff
@@ -43,6 +52,44 @@ class Task extends TypedEventEmitter<ITaskEvents> implements ITask {
...
+            console.log(this.name.getText());
+            this._name.replaceText(0, oldString.length, this.diffName);
+            console.log(this.name.getText());
+        }
+        this.diffPriority = this.DEFAULT_PRIORITY;
+        this.diffType = "none";
+        this.emit('externalPriorityChanged');
+        this.diffName = '';
+        this.diffType = "none";
+        this.emit('externalNameChanged');
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/src/model/taskList.ts:84]: Should we only emit the events when the specific fields changed? I.e. inside the ifs above.

File: examples/hosts/app-integration/external-data/src/model/taskList.ts [hunk 16]
```diff
@@ -43,6 +52,44 @@ class Task extends TypedEventEmitter<ITaskEvents> implements ITask {
...
+    }
+    public ignoreChange = async (): Promise<void> => {
+        this.diffPriority = this.DEFAULT_PRIORITY;
+        this.diffType = "none";
+        this.emit('externalPriorityChanged');
+        this.diffName = '';
+        this.diffType = "none";
+        this.emit('externalNameChanged');
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/src/model/taskList.ts:92]: Emitting here feels weird. If we're ignoring the external changes I don't think these should happen. If anything, they should emit when we get the notification (or poll for) that the external data source had changes?

File: examples/hosts/app-integration/external-data/src/model-interface/index.ts (hunk 1 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/model-interface/index.ts]: Got it! Updated! Thanks for all the links and explanations.

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx [hunk 17]
```diff
@@ -5,17 +5,14 @@
...
 import React, { useEffect, useState } from "react";
 import isEqual from 'lodash.isequal'
-import type { IAppModel, TaskData } from "../model-interface";
+import type { TaskData } from "../model-interface";
 import { customerServicePort } from "../mock-service-interface";
 
 /**
  * {@link DebugView} input props.
  */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/src/view/debugView.tsx:14]: What's the need for the empty interfaces?

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx [hunk 18]
```diff
@@ -158,14 +156,147 @@ function debugResetExternalData(): void {
...
+                    setExternalData(newData);
+                }
+            } catch(error) {
+                console.error(`APP: An error was encountered while polling external data:\n${error}`);
+            }
+        }
+
+        // Run once immediately to run without waiting.
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        pollForServiceUpdates();
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/src/view/debugView.tsx:244]: I see the other case like these updated above, not sure if you saw this one too :). And did it not make the eslint-disable unnecessary? I might have been wrong on that.

File: examples/hosts/app-integration/external-data/src/model/taskList.ts (hunk 16 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/model/taskList.ts:92]: These emits are to trigger re-rending of the React component.

File: examples/hosts/app-integration/external-data/src/model/taskList.ts (hunk 15 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/model/taskList.ts:84]: These emits will trigger removing the UI from the screen.

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx (hunk 17 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/view/debugView.tsx:14]: I suppose we could remove them!

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx (hunk 18 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/view/debugView.tsx:244]: Ah missed it. Let me check about the floating promise - I think it might still be necessary.

File: examples/hosts/app-integration/external-data/src/model-interface/index.ts (hunk 12 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/model-interface/index.ts:106]: I want them to behave like constants, so that's why I put them in all caps. The problem with that is where they should live and how to import them in the right places.  Let me see if I can treat them as proper constants.

File: examples/hosts/app-integration/external-data/src/model/taskList.ts (hunk 14 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/model/taskList.ts:68]: Hah! Interestingly no! I thought they had to be promises because of the events! But I see that's not the case!

File: examples/hosts/app-integration/external-data/src/model/taskList.ts (hunk 15 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/model/taskList.ts]: OH hm I misread that the first time around. I see what you're saying. Let me see.

File: examples/hosts/app-integration/external-data/src/model/taskList.ts [hunk 19]
```diff
@@ -44,6 +48,37 @@ class Task extends TypedEventEmitter<ITaskEvents> implements ITask {
...
         });
     }
+    private readonly resetDiffProperties = (): void => {
+        this.diffPriority = DEFAULT_DIFF_PRIORITY;
+        this.diffType = "none";
+        this.emit('externalPriorityChanged');
+        this.diffName = '';
+        this.diffType = "none";
+        this.emit('externalNameChanged');
+    }
```
Comment (alexvy86) [inline examples/hosts/app-integration/external-data/src/model/taskList.ts:58]: Something about this still feels funky to me but it shouldn't block merging to the dev branch. I understand now the purpose of the events is to cause re-render, but then I wonder if we could design/name the events a bit differently to achieve that. Mainly I still have hard time with events with these names triggering when the user decided to ignore the external changes.

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx (hunk 11 above)
Comment (alexvy86) [reply examples/hosts/app-integration/external-data/src/view/debugView.tsx:184]: Not blocking for the PR, but I think yes? And instead of constructing an `ExternalDataTask` on line 234 you could just return an anonymous object as long as it has the three properties.

**Review by alexvy86 — APPROVED** [review]
No big objections from me; but disclaimer, I'd trust @ChumpChief or @Josmithr 's comments more than my own 😄 .

File: examples/hosts/app-integration/external-data/src/model-interface/index.ts [hunk 20]
```diff
@@ -66,6 +76,34 @@ export interface ITask extends IEventProvider<ITaskEvents> {
      * The task priority.  Modifications are persisted in Fluid and shared amongst collaborators.
      */
     priority: number;
+    /**
+     * The task name coming in from the external server.
+     */
+    diffName: string;
```
Comment (ChumpChief) [inline examples/hosts/app-integration/external-data/src/model-interface/index.ts:82]: Maybe see if there's a more revealing name, maybe sourceName etc.

File: examples/hosts/app-integration/external-data/src/model/containerCode.ts [hunk 21]
```diff
@@ -62,11 +62,11 @@ export class TaskListContainerRuntimeFactory extends ModelContainerRuntimeFactor
         );
         // Register listener only once the model is fully loaded and ready
         runtime.on("signal", (message) => {
-            if (message.type === SignalType.ExternalDataChanged) {
-                // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+        if (message?.content?.type === SignalType.ExternalDataChanged) {        // eslint-disable-next-line @typescript-eslint/no-floating-promises
                 taskList.importExternalData();
```
Comment (ChumpChief) [inline examples/hosts/app-integration/external-data/src/model/containerCode.ts:67]: Recommend adding a `.catch(console.error)` at least rather than floating the promise

File: examples/hosts/app-integration/external-data/src/model-interface/index.ts [hunk 22]
```diff
@@ -30,25 +30,35 @@ export interface IAppModel extends IEventProvider<IAppModelEvents> {
...
-    (event: "nameChanged" | "priorityChanged", listener: () => void);
+    (event: "externalNameChanged", listener: () => void);
+
+    /**
+     * Emitted when extneralPriority has changed.
+     */
+    (event: "externalPriorityChanged", listener: () => void);
 }
 
+export const DEFAULT_DIFF_PRIORITY: number = Number.POSITIVE_INFINITY;
```
Comment (ChumpChief) [inline examples/hosts/app-integration/external-data/src/model-interface/index.ts:58]: Recommend keeping the interface directory as pure interfaces, so moving any consts out to where the implementation will use them.

File: examples/hosts/app-integration/external-data/src/model/taskList.ts (hunk 19 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/model/taskList.ts:58]: Perhaps a better name for them is actually "diffTypeChanged" and "diffNameChanged" because they do change at each trigger of these, once to populate them and render on screen and if ignoring them, then to clear those values and re-render the clearing on screen. So I think I need to do a passthrough to clean up the names and match them to their event triggers.

File: examples/hosts/app-integration/external-data/src/model/taskList.ts [hunk 23]
```diff
@@ -34,7 +35,10 @@ class Task extends TypedEventEmitter<ITaskEvents> implements ITask {
     public constructor(
         private readonly _id: string,
         private readonly _name: SharedString,
-        private readonly _priority: ISharedCell<number>
+        private readonly _priority: ISharedCell<number>,
+        public diffName: string = DEFAULT_DIFF_NAME,
```
Comment (ChumpChief) [inline examples/hosts/app-integration/external-data/src/model/taskList.ts:39]: Recommend using accessors rather than making these directly public, same as the _id/_name.  Also consider making them nullable (`string | undefined`) rather than having to use a default for local-only tasks.

File: examples/hosts/app-integration/external-data/src/model/taskList.ts (hunk 19 above)
Comment (ChumpChief) [reply examples/hosts/app-integration/external-data/src/model/taskList.ts:58]: Couple options:
1. With the naming suggestion above, these could become something like `setSourceName()`
2. Alternatively, if using a setter accessor then the events can be raised from those setters.

File: examples/hosts/app-integration/external-data/src/model/taskList.ts [hunk 24]
```diff
@@ -44,6 +48,37 @@ class Task extends TypedEventEmitter<ITaskEvents> implements ITask {
...
+        this.diffName = savedName;
+        this.diffType = "change";
+        this.emit('externalNameChanged');
+    }
+    public externalPriorityChanged = (savedPriority: number): void => {
+        this.diffPriority = savedPriority;
+        this.diffType = "change";
+        this.emit('externalPriorityChanged');
+    }
+    public acceptChange = (): void => {
```
Comment (ChumpChief) [inline examples/hosts/app-integration/external-data/src/model/taskList.ts:69]: Maybe use a more verbose name just to make it clear what's happening - something like overwriteWithSourceData or something

File: examples/hosts/app-integration/external-data/src/model/taskList.ts [hunk 25]
```diff
@@ -44,6 +48,37 @@ class Task extends TypedEventEmitter<ITaskEvents> implements ITask {
...
+        if (this.diffPriority !== DEFAULT_DIFF_PRIORITY) {
+            this._priority.set(this.diffPriority);
+        }
+        if (this.diffName !== '') {
+            const oldString = this._name.getText()
+            this._name.replaceText(0, oldString.length, this.diffName);
+        }
+        this.resetDiffProperties();
+    }
+    public ignoreChange = (): void => {
```
Comment (ChumpChief) [inline examples/hosts/app-integration/external-data/src/model/taskList.ts:79]: I think ignore probably shouldn't be an option - i.e. it's ok to delay accepting, but ultimately the goal is to converge.  Maybe eventually this becomes something like saveToSource (try to write the fluid data back to the source) but not needed for now.

File: examples/hosts/app-integration/external-data/src/model/taskList.ts [hunk 26]
```diff
@@ -44,6 +48,37 @@ class Task extends TypedEventEmitter<ITaskEvents> implements ITask {
             this.emit("priorityChanged");
         });
     }
+    private readonly resetDiffProperties = (): void => {
+        this.diffPriority = DEFAULT_DIFF_PRIORITY;
```
Comment (ChumpChief) [inline examples/hosts/app-integration/external-data/src/model/taskList.ts:52]: if these are sourceName and sourcePriority, then there's no need to reset them (they will remain truthfully reporting the source values).

File: examples/hosts/app-integration/external-data/src/model/taskList.ts [hunk 27]
```diff
@@ -224,33 +247,44 @@ export class TaskList extends DataObject implements ITaskList {
...
         // TODO: Delete any items that are in the root but missing from the external data
         const updateTaskPs = updatedExternalData.map(async ([id, { name, priority }]) => {
-            const currentTask = this.draftData.get<PersistedTask>(id);
             // Write external data into savedData map.
-            this.savedData.set(id, currentTask);
-
+            const currentTask = this.savedData.get<PersistedTask>(id);
+            // Create a new task because it doesn't exist already
             if (currentTask === undefined) {
+                const savedNameString = SharedString.create(this.runtime);
```
Comment (ChumpChief) [inline examples/hosts/app-integration/external-data/src/model/taskList.ts:254]: Option for a future PR (not this one) -- the saved data doesn't need to be a sharedstring/sharedcell since we're not permitting live edits on it.  We could save the data directly into the savedData map.

File: examples/hosts/app-integration/external-data/src/model/taskList.ts [hunk 28]
```diff
@@ -224,33 +247,44 @@ export class TaskList extends DataObject implements ITaskList {
...
             }
-            const [currName, currPriority] = await Promise.all([
-                currentTask.name.get(),
-                currentTask.priority.get(),
-            ]);
-            if (currName.getText() !== name) {
-                // TODO: Currently replacing existing Fluid data.  But eventually this is where
-                // we'd want conflict resolution UX.
-                currName.replaceText(0, currName.getLength(), name);
+            if (task.name.getText() !== name) {
```
Comment (ChumpChief) [inline examples/hosts/app-integration/external-data/src/model/taskList.ts:281]: Instead of comparing the name here, perhaps it would be better to check above (while writing the new saved data in) whether the new saved data is different from the old saved data above (a change was made remotely) and then additionally whether the current Fluid name is different from the old saved data (a change was made locally).  We only have to negotiate the diff if both of these have happened, otherwise we should be able to resolve silently.

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx [hunk 29]
```diff
@@ -122,20 +109,16 @@ interface ISyncStatusViewProps { }
 const SyncStatusView: React.FC<ISyncStatusViewProps> = (props: ISyncStatusViewProps) => {
     return (
         <div>
-            <h3>Sync status</h3>
+            {/* <h3>Sync status</h3>
```
Comment (ChumpChief) [inline examples/hosts/app-integration/external-data/src/view/debugView.tsx:112]: Probably keep this?  @WayneFerrao has been working in this space so I think he would have to add it back.

File: experimental/framework/react-inputs/src/CollaborativeInput.tsx (hunk 5 above)
Comment (ChumpChief) [reply experimental/framework/react-inputs/src/CollaborativeInput.tsx:60]: Filtering out local events makes sense if the CollaborativeInput is the only one editing the SharedString.  But in this demo, the SharedString is edited by both the CollaborativeInput as well as programmatically, breaking that assumption.

I don't remember if this filtering was just to minimize re-renders vs. if there was a functional aspect to it.  Probably make sure to try a few combinations of making edits in multiple tabs to verify there's no unintended functional change.

**Review by ChumpChief — APPROVED** [review]
Some suggestions but approving since this is a dev branch, no need to block if that's more efficient.

File: examples/hosts/app-integration/external-data/src/model/taskList.ts (hunk 23 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/model/taskList.ts:39]: I tried using (string | undefined) but the react component useEffect threw a real hissy fit about it so I never managed to figure out that pattern. Do you have examples of that that I can copy?

File: examples/hosts/app-integration/external-data/src/model/taskList.ts (hunk 24 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/model/taskList.ts:69]: Let me push up my latest changes with what I think are clearer names.

File: examples/hosts/app-integration/external-data/src/model/taskList.ts (hunk 23 above)
Comment (ChumpChief) [reply examples/hosts/app-integration/external-data/src/model/taskList.ts]: What sort of error?  You'll need to typecheck them before using them as string/number but that's the only thing I would expect.

File: examples/hosts/app-integration/external-data/src/view/taskListView.tsx (hunk 7 above)
Comment (Josmithr) [reply examples/hosts/app-integration/external-data/src/view/taskListView.tsx]: Yup! Technically I think we also want to include the corresponding `set` functions as deps too (since we use them in the effect), but since they will strictly change when their corresponding values change, I think this is safe.

File: examples/hosts/app-integration/external-data/src/model-interface/index.ts [hunk 30]
```diff
@@ -30,25 +30,35 @@ export interface IAppModel extends IEventProvider<IAppModelEvents> {
      * A task tracker list.
      */
     readonly taskList: ITaskList;
-
-    /**
-    * Send custom signals to the server which will cause the server to respond
-    * with the (currently experimental) RuntimeMessage Signal to communicate
-    * an external data change and and possibly the changed data as well
-    */
-    readonly debugSendCustomSignal: () => void;
```
Comment (Josmithr) [inline examples/hosts/app-integration/external-data/src/model-interface/index.ts:39]: Woo! Removing this is a big milestone!

File: examples/hosts/app-integration/external-data/src/model/taskList.ts [hunk 31]
```diff
@@ -34,7 +35,10 @@ class Task extends TypedEventEmitter<ITaskEvents> implements ITask {
     public constructor(
         private readonly _id: string,
         private readonly _name: SharedString,
-        private readonly _priority: ISharedCell<number>
+        private readonly _priority: ISharedCell<number>,
+        public incomingName: string = NONE_INCOMING_NAME,
```
Comment (Josmithr) [inline examples/hosts/app-integration/external-data/src/model/taskList.ts:39]: Should these new properties be externally editable? Or should they be public get, private set?

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx [hunk 32]
```diff
@@ -157,15 +140,147 @@ function debugResetExternalData(): void {
...
+/**
+ * A tabular, editable view of the task list.  Includes a save button to sync the changes back to the data source.
+ */
+export const TaskListView: React.FC = () => {
+    const [externalData, setExternalData] = useState({});
+    useEffect(() => {
+        // HACK: Once we have external changes triggering the appropriate Fluid signal, we can simply listen
+        // for changes coming into the model that way.
+        // For now, poll the external service directly for any updates and apply as needed.
+        async function pollForServiceUpdates(): Promise<void> {
```
Comment (Josmithr) [inline examples/hosts/app-integration/external-data/src/view/debugView.tsx:203]: Can we dedupe this logic with ExternalDataView? Maybe pull it out into a helper function they both use?

File: examples/hosts/app-integration/external-data/src/model/taskList.ts (hunk 23 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/model/taskList.ts]: I think I was getting tripped up with the typecheck earlier. It used the accesser pattern now and it works great!

File: examples/hosts/app-integration/external-data/src/model/taskList.ts (hunk 31 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/model/taskList.ts]: yep changed to getter/setter pattern

File: examples/hosts/app-integration/external-data/src/model/taskList.ts (hunk 28 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/model/taskList.ts:292]: If both these cases haven't happened, do we still want to flag that the data has come in overwritten what was there visually to the client? Or should it be silent in the sense that they don't even know that it has happened?

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx (hunk 11 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/view/debugView.tsx]: All righty! Got it! Changed it! It works!

File: examples/hosts/app-integration/external-data/src/model/taskList.ts (hunk 16 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/model/taskList.ts]: Moved to getter setter pattern and that definitely cleans up this mess of triggers!

File: examples/hosts/app-integration/external-data/src/view/debugView.tsx (hunk 32 above)
Comment (ghoshkaj) [reply examples/hosts/app-integration/external-data/src/view/debugView.tsx]: I've pulled it out into a helper function but I'm not sure the helper function should live! Right now I'm leaving it in DebugView.tsx itself, but let's discuss where to move it out to!
